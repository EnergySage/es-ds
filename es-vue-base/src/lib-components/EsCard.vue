<template>
    <component
        :is="defaultTag"
        :href="href"
        class="es-card"
        :class="actualVariant"
        :to="to"
        v-bind="$attrs"
        v-on="$listeners">
        <slot />
    </component>
</template>

<script lang="js">
export default {
    name: 'EsCard',
    props: {
        href: {
            type: String,
            default: undefined,
        },
        tag: {
            type: String,
            default: 'div',
        },
        // only specified so it's passed through without being set on the resulting HTML element
        to: {
            type: String,
            default: undefined,
        },
        variant: {
            type: String,
            default: 'display',
            required: false,
        },
    },
    computed: {
        actualVariant() {
            // if href is specified, force variant to be interactive
            if (this.href) {
                return 'interactive';
            }

            // if no href provided, honor the specified variant
            return this.variant;
        },
        defaultTag() {
            // if an href is provided and no alternate link tag was provided (e.g. nuxt-link),
            // force the tag to be a b-link, no matter the variant
            if (this.href && this.tag === 'div') {
                return 'b-link';
            }

            // otherwise, honor the tag passed in if it's display variant
            // or something other than the default tag was provided
            if (this.variant !== 'interactive' || this.tag !== 'div') {
                return this.tag;
            }

            // if we reach this point, variant is interactive, tag is div, and no href was provided
            // and any non-link clickable element should be a button
            return 'button';
        },
    },
};
</script>
