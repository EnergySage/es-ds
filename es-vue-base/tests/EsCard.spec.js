import { mount } from '@vue/test-utils';
import EsCard from '@/src/lib-components/EsCard.vue';
import jestVue from '@/tests/jest.vue.config';

describe('EsCard', () => {
    // Basic test; should exist for most components
    test('<EsCard />', async () => {
        const wrapper = mount(EsCard, {
            ...jestVue,
            slots: {
                default: 'Card content',
            },
            propsData: {
                id: 'card-id',
            },
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();

        const card = wrapper.findComponent(EsCard);
        expect(card.element.tagName).toBe('DIV');
        expect(card.classes()).toContain('display');
        expect(card.text()).toBe('Card content');
    });

    // Paragraph tag is honored for display variant
    test('EsCard paragraph tag', async () => {
        const wrapper = mount(EsCard, {
            ...jestVue,
            slots: {
                default: 'Card content',
            },
            propsData: {
                id: 'card-id',
                tag: 'p',
            },
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();

        const card = wrapper.findComponent(EsCard);
        expect(card.element.tagName).toBe('P');
        expect(card.classes()).toContain('display');
        expect(card.text()).toBe('Card content');
    });

    // Default interactive variant
    test('EsCard default interactive variant', async () => {
        const wrapper = mount(EsCard, {
            ...jestVue,
            slots: {
                default: 'Card content',
            },
            propsData: {
                id: 'card-id',
                variant: 'interactive',
            },
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();

        const card = wrapper.findComponent(EsCard);
        expect(card.element.tagName).toBe('BUTTON');
        expect(card.classes()).toContain('interactive');
        expect(card.text()).toBe('Card content');
    });

    // Interactive from href, no variant specified
    test('EsCard becomes interactive when href is passed and variant is omitted', async () => {
        const href = 'https://www.energysage.com/';
        const wrapper = mount(EsCard, {
            ...jestVue,
            slots: {
                default: 'Card content',
            },
            propsData: {
                id: 'card-id',
                href,
            },
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();

        const card = wrapper.findComponent(EsCard);
        expect(card.element.tagName).toBe('A');
        expect(card.classes()).toContain('interactive');
        expect(card.attributes().href).toBe(href);
        expect(card.text()).toBe('Card content');
    });

    // Interactive from href, display variant specified
    test('EsCard becomes interactive when href is passed and display variant is passed', async () => {
        const href = 'https://www.energysage.com/';
        const wrapper = mount(EsCard, {
            ...jestVue,
            slots: {
                default: 'Card content',
            },
            propsData: {
                id: 'card-id',
                href,
                variant: 'display', // this should be ignored, due to href
            },
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();

        const card = wrapper.findComponent(EsCard);
        expect(card.element.tagName).toBe('A');
        expect(card.classes()).toContain('interactive');
        expect(card.attributes().href).toBe(href);
        expect(card.text()).toBe('Card content');
    });

    // Interactive from variant and href
    test('EsCard interactive with href', async () => {
        const href = 'https://www.energysage.com/';
        const wrapper = mount(EsCard, {
            ...jestVue,
            slots: {
                default: 'Card content',
            },
            propsData: {
                id: 'card-id',
                href,
                variant: 'interactive',
            },
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();

        const card = wrapper.findComponent(EsCard);
        expect(card.element.tagName).toBe('A');
        expect(card.classes()).toContain('interactive');
        expect(card.attributes().href).toBe(href);
        expect(card.text()).toBe('Card content');
    });

    // Tag is honored when passed to interactive variant with href
    test('EsCard tag is honored when passed to interactive variant with href', async () => {
        const href = 'https://www.energysage.com/';
        const wrapper = mount(EsCard, {
            ...jestVue,
            slots: {
                default: 'Card content',
            },
            propsData: {
                id: 'card-id',
                href,
                tag: 'p', // invalid html with href, but suitable test proxy for nuxt-link
                variant: 'interactive',
            },
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();

        const card = wrapper.findComponent(EsCard);
        expect(card.element.tagName).toBe('P');
        expect(card.classes()).toContain('interactive');
        expect(card.attributes().href).toBe(href);
        expect(card.text()).toBe('Card content');
    });
});
