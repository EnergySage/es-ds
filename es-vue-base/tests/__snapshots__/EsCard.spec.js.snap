// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsCard <EsCard /> 1`] = `"<div id="card-id" class="es-card display">Card content</div>"`;

exports[`EsCard EsCard becomes interactive when href is passed and display variant is passed 1`] = `"<a id="card-id" href="https://www.energysage.com/" target="_self" class="es-card interactive">Card content</a>"`;

exports[`EsCard EsCard becomes interactive when href is passed and variant is omitted 1`] = `"<a id="card-id" href="https://www.energysage.com/" target="_self" class="es-card interactive">Card content</a>"`;

exports[`EsCard EsCard default interactive variant 1`] = `"<button id="card-id" class="es-card interactive">Card content</button>"`;

exports[`EsCard EsCard interactive with href 1`] = `"<a id="card-id" href="https://www.energysage.com/" target="_self" class="es-card interactive">Card content</a>"`;

exports[`EsCard EsCard paragraph tag 1`] = `"<p id="card-id" class="es-card display">Card content</p>"`;

exports[`EsCard EsCard tag is honored when passed to interactive variant with href 1`] = `"<p href="https://www.energysage.com/" id="card-id" class="es-card interactive">Card content</p>"`;
